network-store-client/src/main/java/com/powsybl/network/store/client/AbstractTopology.java

```diff
@@ -245,57 +249,64 @@ protected void buildGraph(NetworkObjectIndex index, Resource<VoltageLevelAttribu
+        return calculatedBusAttributesList;
     }
 
     public Map<String, Bus> calculateBuses(NetworkObjectIndex index, Resource<VoltageLevelAttributes> voltageLevelResource) {
-        Map<T, List<Vertex<T>>> verticesByNodeOrBus = new HashMap<>();
-        UndirectedGraph<T, Resource<SwitchAttributes>> graph = buildGraph(index, voltageLevelResource, verticesByNodeOrBus);
-        return calculateBuses(index, voltageLevelResource, graph, verticesByNodeOrBus, AbstractTopology::busViewBusValidator);
-    }
+        List<CalculatedBusAttributes> calculatedBusAttributesList = getCalculatedBusAttributesList(index, voltageLevelResource);
 
-    private CalculatedBus<T> calculateBuses(NetworkObjectIndex index,
-                                            Resource<VoltageLevelAttributes> voltageLevelResource,
-                                            UndirectedGraph<T, Resource<SwitchAttributes>> graph,
-                                            Map<T, List<Vertex<T>>> verticesByNodeOrBus,
-                                            BiPredicate<Map<T, List<Vertex<T>>>, Set<T>> busValidator,
-                                            T startNodeOrBus) {
-        Set<T> nodesOrBuses = new ConnectivityInspector<>(graph).connectedSetOf(startNodeOrBus);
-        return tryToCreateCalculatedBus(index, voltageLevelResource, verticesByNodeOrBus, nodesOrBuses, busValidator).orElse(null);
+        Map<String, Bus> calculatedBuses = new HashMap<>(calculatedBusAttributesList.size());
+        for (int calculatedBusNum = 0; calculatedBusNum < calculatedBusAttributesList.size(); calculatedBusNum++) {
+            CalculatedBus calculatedBus = createCalculatedBus(index, voltageLevelResource, calculatedBusNum);
+            calculatedBuses.put(calculatedBus.getId(), calculatedBus);
+        }
+        return calculatedBuses;
     }
 
-    public CalculatedBus<T> calculateBuses(NetworkObjectIndex index, Resource<VoltageLevelAttributes> voltageLevelResource, T startNodeOrBus) {
-        Map<T, List<Vertex<T>>> verticesByNodeOrBus = new HashMap<>();
-        UndirectedGraph<T, Resource<SwitchAttributes>> graph = buildGraph(index, voltageLevelResource, verticesByNodeOrBus);
-        return calculateBuses(index, voltageLevelResource, graph, verticesByNodeOrBus, AbstractTopology::busViewBusValidator, startNodeOrBus);
+    public CalculatedBus calculateBus(NetworkObjectIndex index, Resource<VoltageLevelAttributes> voltageLevelResource, T nodeOrBus) {
+        List<CalculatedBusAttributes> calculatedBusAttributesList = getCalculatedBusAttributesList(index, voltageLevelResource);
+
+        for (int calculatedBusNum = 0; calculatedBusNum < calculatedBusAttributesList.size(); calculatedBusNum++) {
+            CalculatedBusAttributes calculatedBusAttributes = calculatedBusAttributesList.get(calculatedBusNum);
+            for (Vertex vertex : calculatedBusAttributes.getVertices()) {
+                if (getNodeOrBus(vertex).equals(nodeOrBus)) {
```

Comment:
should we store a map of vertices to bus to speed up this lookup or is this always small ?